[MW-225] feat(auth): Implement full Sign In & Sign Up flow with validation

HekmatullahAmin opened the pull request:
Jira Task: [MW-225](https://mifosforge.jira.com/browse/MW-225)

## Screenshots


https://github.com/user-attachments/assets/29712344-cd42-4379-a8d6-6ad8341ef83e


## Description
### What’s Achieved in This PR
1. Implemented full Sign In and Sign Up flow

- Covers both happy and unhappy user journeys.

- Seamlessly transitions from input to account creation and navigation.

2. Added robust user input validations

- Ensures mandatory fields like name, email, and mobile number are filled.

- Email is checked for correct format.

- Password and confirm password must match.

3. Improved error handling

- All validation and API errors are now clearly presented to the user.

- Replaced generic or null messages with user-friendly alternatives.

- Handles unique field validation (username and mobile) with clear feedback.

4. Password requirement hinting

- A help card shows password rules inline before submission.

- Prevents trial-and-error by guiding users upfront.

5. Updated pin code validation

- Removed strict 6-digit requirement for pin codes.

- Allows shorter lengths to support different country formats (e.g., 4- or 5-digit codes).

6. Country and state dropdown integration

- Integrated [stefanbinder/countries-states](https://github.com/stefanbinder/countries-states) JSON for populating country-state mapping.

- Countries and their respective states are loaded from countries.json and rendered as dropdowns in the signup form.

##
<!--Please make sure these boxes are checked before submitting your pull request - thanks!-->

- [ ] Apply the `AndroidStyle.xml` style template to your code in Android Studio.

- [ ] Run the unit tests with `./gradlew check` to make sure you didn't break anything

- [ ] If you have multiple commits please combine them into one commit by squashing them.


[MW-225]: https://mifosforge.jira.com/browse/MW-225?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

HekmatullahAmin committed:
Add complete sign-in and sign-up flow with input validation and visible error messages for all user actions

HekmatullahAmin committed:
Add QuestionMark icon

HekmatullahAmin committed:
remove the logs

HekmatullahAmin committed:
fix static analysis check

niyajali reviewed (changes requested):
@HekmatullahAmin Fetch Country & State from server and show as drop-down if any API available otherwise create a list of country and state and show as drop-down.

And instead of dialog for password feedback show a card below of the password indicator which should contain list of password feedback.

niyajali commented:
And update the PR title like [MW-225] - The PR Title

HekmatullahAmin committed:
add password requirements card and country-state dropdown support

niyajali commented on feature/auth/src/commonMain/kotlin/org/mifospay/feature/auth/signup/SignupScreen.kt:
```diff
@@ -95,6 +111,10 @@ internal fun SignupScreen(
         }
     }
 
+    LaunchedEffect(Unit) {
```
remove this and call inside of `init` block on ViewModel

niyajali commented on feature/auth/src/commonMain/kotlin/org/mifospay/feature/auth/signup/SignupScreen.kt:
```diff
@@ -317,25 +358,65 @@ private fun SignupScreenContent(
                     onValueChange = {
                         onAction(SignUpAction.CountryInputChange(it))
+                        // Clear state when country changes
+                        onAction(SignUpAction.StateInputChange(""))
```
You can update in ViewModel, no need to send multiple action

niyajali commented on feature/auth/src/commonMain/kotlin/org/mifospay/feature/auth/signup/Country.kt:
```diff
@@ -0,0 +1,23 @@
+
+@Serializable
+internal data class Country(
+    val name: String,
```
add `countryCode` parameter

niyajali commented on feature/auth/src/commonMain/kotlin/org/mifospay/feature/auth/signup/SignupViewModel.kt:
```diff
@@ -151,7 +153,11 @@ class SignupViewModel(
+                    it.copy(
+                        countryInput = action.country,
+                        // reset state when country changes
+                        stateInput = "",
```
see here you reseting the state, so you no need to send multiple action from UI

niyajali commented on feature/auth/src/commonMain/kotlin/org/mifospay/feature/auth/signup/SignupViewModel.kt:
```diff
@@ -509,6 +504,33 @@ class SignupViewModel(
+                    // This will skip unexpected fields
+                    ignoreUnknownKeys = true
+                }
+                val bytes = Res.readBytes("files/countries.json")
```
Move this file to the data module and create a interface like `AssetRepository or LocalAssetRepository` and create required methods to load this data so you can use it in multiple places.

niyajali commented on feature/auth/src/commonMain/kotlin/org/mifospay/feature/auth/signup/SignupViewModel.kt:
```diff
@@ -529,27 +551,9 @@ data class SignUpState(
-    val isPasswordMatch: Boolean
-        get() = passwordInput == confirmPasswordInput
-}
+    val passwordFeedback: List<String> = emptyList(),
```
Use `ImmutableList<T>` instead of `List<T>`, which is not stable for better performance. Add the `kotlinx.collection.immutable` and `kotlinx.serialization.json` dependencies to the KMP convention plugin to make them accessible across all modules.

HekmatullahAmin committed:
Refactor countries.json handling with AssetRepository and persistent state support in signup flow

niyajali commented on core/designsystem/src/commonMain/kotlin/org/mifospay/core/designsystem/component/TextField.kt:
```diff
@@ -74,9 +74,10 @@ fun MifosOutlinedTextField(
         readOnly = readOnly,
         supportingText = {
-            if (isError) {
+            if (isError && isFocused) {
```
Why we're checking textfeld is focused or not, the user has to click on each text field to see the error

HekmatullahAmin replied:
yeah i put it like that because if we don't then when user opens sign up screen all the fields will be in red. But they eventually go to each field even if they press a dialog will pop up to enter it. I saw some mater design the red appear after clicking or typing in the `TextField` so cause of that i did it like this.

niyajali commented on core/designsystem/src/commonMain/kotlin/org/mifospay/core/designsystem/component/TextField.kt:
```diff
@@ -168,10 +169,10 @@ fun MifosTextField(
         placeholder = placeholder,
-        supportingText = errorText?.let {
-            {
+        supportingText = {
```
do like this, so it doesn't take space for `supportingText` field, otherwise this(`if (isError && isFocused && errorText != null) ` )  may be true but the `supportingText ` composable will intitalize initially
```
errorText?.let {
    { }
 }

HekmatullahAmin replied:
ok so you mean use `errorText?.let` and then inside it do the checking.

niyajali commented on core/ui/src/commonMain/kotlin/org/mifospay/core/ui/MifosPasswordField.kt:
```diff
@@ -91,9 +97,9 @@ fun MifosPasswordField(
                 ),
             ) {
                 val imageVector = if (showPassword) {
-                    MifosIcons.OutlinedVisibilityOff
```
`MifosIcons.OutlinedVisibilityOff` should be prefered when we're showing the password

HekmatullahAmin replied:
I swap it because:
- when the eye is open -> we can't see the password
- when the eye was shut (close) -> we can see the password it was vice versa.

you can have a look at the video when the password is not visible the eye is open, and when the password is visible the eye is shot.
So i think when the eye is shut it means you can't see. (Samarth pointed this out and I think it should be like this, what do u think?)

niyajali commented on feature/auth/src/commonMain/kotlin/org/mifospay/feature/auth/signup/SignupScreen.kt:
```diff
@@ -231,12 +250,15 @@ private fun SignupScreenContent(
                 onValueChange = {
                     onAction(SignUpAction.MobileNumberInputChange(it))
                 },
+                errorMessage = stringResource(Res.string.feature_auth_error_mobile_empty),
```
As we're showing the error messages in a dialog, adding this additional message is not necessary or required.

HekmatullahAmin replied:
yeah you are right the user will see those on dialog, but that is after clicking the submit. So before submitting i wanted to user know that putting it empty is not acceptable. wanted to providing real-time feedback

HekmatullahAmin commented:
@niyajali just wanted to clarify the UX expectations here:

Would you prefer:

- Show red outline + support text even when the field hasn't been focused yet (i.e. as soon as the screen loads),
- Show only a `*required` message below each field (no red border or red support text styling),
- Or some other validation behavior?

<img width="644" alt="image" src="https://github.com/user-attachments/assets/2f970e2d-2e60-4f35-967b-2c3d6a80791a" />

niyajali commented:
@HekmatullahAmin What type of validation we had earlier

niyajali commented:
@HekmatullahAmin Okay so currently you're passing some errorMessage on the text field initially, and what is that message, and earlier we're validating the input and showing the error in a dialog on click of submit button

HekmatullahAmin commented:
> @HekmatullahAmin Okay so currently you're passing some errorMessage on the text field initially, and what is that message

yes the message are all like this field cannot be empty. 
for example `username is required `
`address line 1 connot be empty`
...

and will only show if they are empty and is focusable, so if the go to another field it shouldn't show. if not focusable and the user visit the sign up screen for the first time all of those fields is going to look red with their error message.

do you want me to keep it or revert it back to before or u want sth else?

HekmatullahAmin committed:
Removed error message from fields

HekmatullahAmin commented:
@niyajali I revert it and also removed the isError from composables in SignUp Screen. Previously it was passed from the SignUp screen checking if those were empty but in `MifosTextField` the isError was not passed. 
- So as previously there were no red error container nor error message i removed them

biplab1 commented on core/data/src/commonMain/kotlin/org/mifospay/core/data/repositoryImp/AssetRepositoryImpl.kt:
```diff
@@ -0,0 +1,36 @@
+ *
+ * See https://github.com/openMF/mobile-wallet/blob/master/LICENSE.md
+ */
+package org.mifospay.core.data.repositoryImp
```
There is a typo in repositoryImp, it should be repositoryImpl (also in directory name)

biplab1 commented on core/data/src/commonMain/kotlin/org/mifospay/core/data/repositoryImp/AssetRepositoryImpl.kt:
```diff
@@ -0,0 +1,36 @@
+            val json = Json { ignoreUnknownKeys = true }
+            val bytes = Res.readBytes("files/countries.json")
+            val jsonString = bytes.decodeToString()
+            val parsed = json.decodeFromString<List<Country>>(jsonString)
```
`parsed` could be renamed as `countries`

biplab1 commented on core/data/src/commonMain/kotlin/org/mifospay/core/data/repositoryImp/AssetRepositoryImpl.kt:
```diff
@@ -0,0 +1,36 @@
+            val parsed = json.decodeFromString<List<Country>>(jsonString)
+            DataState.Success(
+                parsed.associate { country ->
+                    country.name to country.states.map { it.name }.ifEmpty { listOf("N/A") }
```
The repository should return the raw list (even if it’s empty) and let the UI decide how to present that to the user.

HekmatullahAmin replied:
Yes, it would be a good idea to handle empty state lists in the UI if we wanted to show a different UI or a specific message for the empty case. But for now, our logic is simple—we just want to display a list of states for the selected country. If a country doesn't have states, we show “N/A.” Handling this in the repository keeps our UI cleaner by avoiding conditional checks there.

HekmatullahAmin replied:
@biplab1 i handled the empty case scenario in UI casue now we will use a `TextField` if no state is available so user enter their own, otherwise it will list states from our static data in dropdown

HekmatullahAmin committed:
fixed typos

HekmatullahAmin commented:
@biplab1 could u please resolve your conflicts and approve

revanthkumarJ commented on feature/auth/src/commonMain/kotlin/org/mifospay/feature/auth/signup/SignupViewModel.kt:
```diff
@@ -285,44 +295,38 @@ class SignupViewModel(
                     dialogState = SignUpDialog.Error(
-                        "Password must be at least $MIN_PASSWORD_LENGTH characters long.",
+                        message = "The password field cannot be empty.",
                     ),
```
won't we are using stringResources instead of directly using strings

HekmatullahAmin replied:
yeah it can be done i will do it just need to take care because then our error will who two types of message one `StringResource` and other `String/

HekmatullahAmin replied:
@revanthkumarJ done created a utility functino to take care of it and the user can pass `string` or `stringResource` and the dialog will display the message

itsPronay commented on feature/auth/src/commonMain/kotlin/org/mifospay/feature/auth/signup/SignupScreen.kt:
```diff
@@ -317,25 +318,51 @@ private fun SignupScreenContent(
             ) {
-                MifosOutlinedTextField(
-                    value = state.countryInput,
+                var isCountryDropdownExpanded by remember { mutableStateOf(false) }
+                var isStateDropdownExpanded by remember { mutableStateOf(false) }
+
+                val selectedCountry = state.countryInput
+                val selectedState = state.stateInput
+                val statesForSelectedCountry =
+                    state.countriesWithStates[selectedCountry] ?: emptyList()
+
```
shouldn't we include them inside the `SignUpState` and access them from there!?!

HekmatullahAmin replied:
yes that make sense i will do it

HekmatullahAmin replied:
@itsPronay done

revanthkumarJ commented on core/model/src/commonMain/kotlin/org/mifospay/core/model/utils/Country.kt:
```diff
@@ -0,0 +1,29 @@
+    val code2: String,
+    val code3: String,
+    val name: String,
+    val capital: String,
```
i think it's better to add default values as in api result if any arg not found it will throw an error but if we add default values it won't throw error.

HekmatullahAmin replied:
the data is static and those field is always present so casue of that no need to defaults.

HekmatullahAmin committed:
Add DialogManager utility for global dialog state handling

HekmatullahAmin committed:
fix format violation

HekmatullahAmin commented:
@itsPronay , @revanthkumarJ , @biplab1 when u are available have a look and resolve your conversation and kindly approve

biplab1 commented on core/ui/src/commonMain/kotlin/org/mifospay/core/ui/PasswordStrengthIndicator.kt:
```diff
@@ -21,26 +23,246 @@ import androidx.compose.foundation.layout.Spacer
+                                Text(
+                                    text = strengthLabel,
+                                    style = MaterialTheme.typography.labelSmall,
+                                    color = Color.White,
```
Would it make sense to use `MaterialTheme.colorScheme` here instead of hardcoding the color? Might help with consistency and theming

HekmatullahAmin replied:
done

biplab1 commented on core/ui/src/commonMain/kotlin/org/mifospay/core/ui/utils/PasswordChecker.kt:
```diff
@@ -83,14 +74,11 @@ object PasswordChecker {
         if (password.hasConsecutiveRepetitions()) {
-            feedback.add("Remove consecutive repeating characters.")
+            feedback.add("Avoid using consecutive repeated characters.")
         }
         if (password.hasSpaces()) {
-            feedback.add("Remove spaces.")
+            feedback.add("Do not include spaces in the password.")
```
Since we’re targeting a global audience with our mobile wallet, wouldn’t it be better to use string resources instead of hardcoded strings? I noticed this usage of hardcoded strings in several places. Using `strings.xml` would make it much easier to manage translations later on.

HekmatullahAmin replied:
@biplab1 for now, I'm not moving the password feedback messages to `string.xml`. The reason is that `getPasswordFeedback()` is called from the `SignUpViewModel`, and the result is stored in `SignUpState`. Since SignUpState is marked as `@Parcelize` (for SavedStateHandle support), it can't contain a List<StringResource>.

Also, I can't use `stringResource()` inside getPasswordFeedback() because it's not a composable function, and stringResource() is only accessible from composables.

We'll revisit this later once we refactor how localized strings are handled in view models. for now added a TODO

biplab1 replied:
Makes sense given the current limitations with `@Parcelize` and not being able to access string resources outside composables. 

Just sharing — I’ve used `org.jetbrains.compose.resources.getString` inside a coroutine in one of my modules to get localized strings, and it worked fine. Might be something we can explore during the refactor. 

Reference: [To get string resources as a String from non-composable code](https://www.jetbrains.com/help/kotlin-multiplatform-dev/compose-multiplatform-resources-usage.html#p91qgw_112)

HekmatullahAmin committed:
Added string resources
